src/sinol_make/commands/run/__init__.py

```diff
@@ -360,78 +367,190 @@ def print_view(output_file=None):
+		self.print_expected_scores(expected_scores)
+		print(util.bold("\nExpected scores based on results:"))
+		self.print_expected_scores(new_expected_scores)
+
+		expected_scores_diff = dictdiffer.diff(expected_scores, new_expected_scores)
+		added_solutions = set()
+		removed_solutions = set()
+		added_groups = set()
+		removed_groups = set()
+		changes = [] # Array of tuples (solution, group, current_result, expected_result)
+
+		for type, field, change in list(expected_scores_diff):
+			if type == "add":
+				if field == '': # Solutions were added
+					for solution in change:
+						added_solutions.add(solution[0])
+				elif field[1] == "expected": # Groups were added
+					for group in change:
+						added_groups.add(group[0])
+			elif type == "remove":
+				# We check whether a solution was removed only when sinol_make was run on all of them
+				if field == '' and self.args.programs == None and config_expected_scores:
+					for solution in change:
+						removed_solutions.add(solution[0])
+				# We check whether a group was removed only when sinol_make was run on all of them
+				elif field[1] == "expected" and self.args.tests == None and config_expected_scores:
+					for group in change:
+						removed_groups.add(group[0])
+			elif type == "change":
+				if field[1] == "expected": # Results for at least one group has changed
+					solution = field[0]
+					group = field[2]
+					old_result = change[0]
+					result = change[1]
+					changes.append((solution, group, result, old_result))
+
+		return {
+			"added_solutions": added_solutions,
+			"removed_solutions": removed_solutions,
+			"added_groups": added_groups,
+			"removed_groups": removed_groups,
+			"changes": changes,
+			"expected_scores": expected_scores,
+			"new_expected_scores": new_expected_scores,
+		}
+
+
+	def print_expected_scores_diff(self, validation_results):
+		added_solutions = validation_results["added_solutions"]
+		removed_solutions = validation_results["removed_solutions"]
+		added_groups = validation_results["added_groups"]
+		removed_groups = validation_results["removed_groups"]
+		changes = validation_results["changes"]
+		expected_scores = validation_results["expected_scores"]
+		new_expected_scores = validation_results["new_expected_scores"]
+		config_expected_scores = self.config["sinol_expected_scores"] if "sinol_expected_scores" in self.config else {}
+
+		def warn_if_not_empty(set, message):
+			if len(set) > 0:
+				print(util.warning(message), end='')
+				print(util.warning(", ".join([str(x) for x in set])))
+
+		warn_if_not_empty(added_solutions, "Solutions were added: ")
+		warn_if_not_empty(removed_solutions, "Solutions were removed: ")
+		warn_if_not_empty(added_groups, "Groups were added: ")
+		warn_if_not_empty(removed_groups, "Groups were removed: ")
+
+		for change in changes:
+			print(util.warning("Solution %s passed group %d with status %s while it should pass with status %s." %
+										(change[0], change[1], change[2], change[3])))
```

Comment:
It's not clear which field contains what. I could move up to see a comment describing the fields, but a better solution would be to name those fields, through e.g. a struct.

Reply:
Changed in 189dd7f695d152db99a0ee7d6ddde26c4e1c1a45